File: leetcode/jump_game_II/JumpGame.cpp
class Solution {
public:
    int jump(int A[], int n) {
        if (n <= 1) {
            return 0; 
        }
        
        int T[n]; 
        for (int i = 0; i < n; i++) {
            T[i] = INT_MAX; 
        }
        
        T[0] = 0; 
        
        // Initialization 
        for (int i = 1; i <= A[0]; i++) {
            T[i] = 1; 
        }
        
        for (int i = A[0] + 1; i < n; i++) {
            int j = i - 1; 
            int min = T[j]; 
            while (j > 0) {
                if (A[j] >= (i - j) && T[j] < min) {
                    min = T[j]; 
                }
                
                j--; 
            }
            
            T[i] = min + 1; 
        }
        
        return T[n - 1]; 
    }
};

File: leetcode6/jump_game.java
/**
 * For a list of max steps, determine whether we can jump from the start to the end. 
 * The idea was to maintian a boolean array meaning whether we can get to a given position. 
 * Initialize true in the array with all the positions we can reach with one jump to save 
 * some computation. For the rest indexs, check every previous reachable position and see 
 * if the current index can be reached from there. 
 */ 
public class Solution {
    public boolean canJump(int[] nums) {
        int len = nums.length; 
        
        if (len == 0) {
            return false; 
        }
        
        boolean[] flag = new boolean[len]; 
        int i = 0; 
        
        while (i < len && i <= nums[0]) {
            flag[i] = true; 
            i++; 
        }
        
        while (i < len) {
            int j = i - 1; 
            
            while (j >= 0) {
                if (flag[j] && nums[j] >= (i - j)) {
                    flag[i] = true; 
                    break; 
                }
                
                j--; 
            }
            
            i++; 
        }
        
        return flag[len - 1]; 
    }
}
